pgjdbc/src/main/java/org/postgresql/core/v3/ScramAuthenticator.java

```diff
@@ -0,0 +1,192 @@
+      ScramClient client = ScramClient.builder()
+          .advertisedMechanisms(advertisedMechanisms)
+          .username("*") // username is ignored by server, startup message is used instead
+          .password(password)
+          .channelBinding(TlsServerEndpoint.TLS_SERVER_END_POINT, cbindData)
+          .stringPreparation(StringPreparation.POSTGRESQL_PREPARATION)
+          .build();
+
+      if (LOGGER.isLoggable(Level.FINEST)) {
+        LOGGER.log(Level.FINEST, " Using SCRAM mechanism: {0}",
+            client.getScramMechanism().getName());
+      }
+      return client;
+    } catch (IllegalArgumentException | IOException e) {
+      throw new PSQLException(
+          GT.tr("Invalid SCRAM client initialization", e),
+          PSQLState.CONNECTION_REJECTED);
+    }
+  }
+
+  private static List<String> advertisedMechanisms(PGStream stream, ChannelBindingOption channelBinding)
+      throws PSQLException, IOException {
+    List<String> mechanisms = new ArrayList<>();
+    do {
+      mechanisms.add(stream.receiveString());
+    } while (stream.peekChar() != 0);
+    int c = stream.receiveChar();
+    assert c == 0;
+    if (mechanisms.isEmpty()) {
+      throw new PSQLException(
+          GT.tr("Received AuthenticationSASL message with 0 mechanisms!"),
+          PSQLState.CONNECTION_REJECTED);
+    }
+    LOGGER.log(Level.FINEST, " <=BE AuthenticationSASL( {0} )", mechanisms);
+    if (channelBinding == ChannelBindingOption.REQUIRE
+        && !mechanisms.stream().anyMatch(m -> m.endsWith("-PLUS"))) {
+      throw new PSQLException(
+          GT.tr("Channel Binding is required, but server did not offer an "
+              + "authentication method that supports channel binding"),
+          PSQLState.CONNECTION_REJECTED);
+    }
+    return mechanisms;
+  }
+
+  private static byte[] getChannelBindingData(PGStream stream, ChannelBindingOption channelBinding)
+      throws PSQLException {
+    if (channelBinding == ChannelBindingOption.DISABLE) {
+      return new byte[0];
+    }
+
+    Socket socket = stream.getSocket();
+    if (socket instanceof SSLSocket) {
+      SSLSession session = ((SSLSocket) socket).getSession();
+      try {
+        Certificate[] certificates = session.getPeerCertificates();
+        if (certificates != null && certificates.length > 0) {
+          Certificate peerCert = certificates[0]; // First certificate is the peer's certificate
+          if (peerCert instanceof X509Certificate) {
+            X509Certificate cert = (X509Certificate) peerCert;
+            return TlsServerEndpoint.getChannelBindingData(cert);
+          }
+        }
+      } catch (CertificateEncodingException | SSLException e) {
+        if (LOGGER.isLoggable(Level.FINEST)) {
+          LOGGER.log(Level.FINEST, "Error extracting channel binding data", e);
+        }
+      }
+    }
+    if (channelBinding == ChannelBindingOption.REQUIRE) {
```

Comment:
This seems confusing to me. If we catch the above exception we could have SSL in use but it failed for some reason. Seems to me this should be executed as an else (socket is not an SSLSocket)

Reply:
Done.